# INFO 2950 Homework 1

This week, our goal is to (re-)familiarize ourselves with the basics of Python, and start getting to know the `pandas` package, which helps us work with data frames. We will also discuss cleaning data. 

## Part 1: Discussion Question

As a data scientist, you'll be asked to build various tools and systems all the time. While you might be *capable* of performing these tasks, it's always worth asking yourself: should you?

Below is a list of scenarios. In each case, answer the discussion question: **If you were to build this system, how might its use affect people (be it positively or negatively)?**

### Scenarios
1. You work for an e-commerce site that wants to direct customers to items they would likely be interested in and avoiding serving them irrelevant ads. You've been asked to build an advertisement system that recommends products to customers based on a variety of attributes, including their age, income level, locale, gender, etc.
1. You work for a news site that lives or dies from ad revenue. Advertisers don't want their ads next to "controversial" news items. You've been asked to build a classifier for "touchy subjects".
1. You work for a public health agency that wants to identify people exposed to a disease for testing. You've been asked to build a system that will trace phone locations and estimate proximity between individuals to infer exposure.
1. You work for a university that wants to identify students having difficulties with stress and mental health. You've been asked to build a system that will identify students who aren't going to class.


---

### Problem 1 (6 pts)
*Remember not to move or modify problem header cells, like this one; we use it for automated homework analysis.* 

Choose **two** of the above scenarios, and in each case, answer the discussion question in approximately 1 paragraph (100 words).

---

(insert answer for one scenario here)

(insert answer for the other scenario here)

---

## Part 2: Python Basics

Twitter is full of controversies, but a recent one over whether or not 2 + 2 = 5 sparked a lot of discussion in the math and stats world. Here's [a great thread](https://twitter.com/kareem_carr/status/1289724475609501697) on this issue.

A key point from this thread: 

_"Our numbers, our quantitative measures, are abstractions of real underlying things in the universe and it's important to keep track of this when we use numbers to model the real world."_

So what does this Twitter controversy have to do with Python...? Let's try to compute 2 + 2 in a few ways and see what happens.


In [1]:
## example 1
x = 2
print("2 + 2 =", x + x)

## example 2
y = 2.0
print("2 + 2 =", y + y)

## example 3
z = "2"
print("2 + 2 =", z + z)

2 + 2 = 4
2 + 2 = 4.0
2 + 2 = 22


2 + 2 = ... 22? What's going on here?

---

### Problem 2 (4 pts)

What is the difference between each of the three examples of arithmetic above? In particular, explain why example 3 produced "2 + 2 = 22". 

---

---

We can perform any number of simple arithmetic operations with Python, but its power lies in our ability automate simple tasks with a few lines of code. In doing so, for loops are often very useful.

---

### Problem 3 (6 pts)

Start with the list of integers from 10 to 20 (inclusive). **Write a `for` loop** that iterates through this list to create a new list, where each entry of the list multiplied by 5 and then increased by 4. Print the final list.

*(**Hint:** use the [`range`](https://python-reference.readthedocs.io/en/latest/docs/functions/range.html?highlight=range) function to iterate over in your for loop.)*

*(**Confidence check:** your code should print `[54, 59, 64, 69, 74, 79, 84, 89, 94, 99, 104]`.)* 

---

In [2]:
## insert answer here

---

For loops are very literal, which makes them straightforward to understand and write. However, they can get a little verbose and they are not always the most computationally efficient.

List comprehensions are a sleek alternative that you will (hopefully) come to love in this course. Their syntax is designed to be simple. It looks like this:

`[`expression `for` item `in` list`]`

For instance, the for loop you wrote in problem 3 can be written succinctly with a list comprehension:

In [7]:
new_list_of_ints = [i*5+4 for i in range(10,21)]
print(new_list_of_ints)

[54, 59, 64, 69, 74, 79, 84, 89, 94, 99, 104]


You can also include `if` conditions on the original list in the list comprehension...

---

### Problem 4 (4 pts)

Copy the above list comprehension and add an if statement (within the list comprehension) after `range(10,21)` so that the list it creates includes only integers that leave a remainder of 3 when divided by 4. The `i*5+4 for i in range(10,21)` part of the list comprehension should remain completely unmodified in your solution.   

*(**Hint:** You may want to make use the [modulus](https://python-reference.readthedocs.io/en/latest/docs/operators/modulus.html) operator)*

*(**Confidence check:** your code should print `[59, 79, 99]`.)* 

---

In [1]:
new_list_of_ints = [i*5+4 for i in range(10,21) if i%4 == 3]
print(new_list_of_ints)

[59, 79, 99]


---

## Part 3: `pandas` and Data Cleaning

The [`pandas`](https://pandas.pydata.org/) package is used to perform data analysis. It is essentially a package for working with data frames (i.e. tables of data).

In [2]:
import pandas as pd

We could simply write `import pandas`, but then every time we want to refer to a function in the `pandas` package (e.g. one called `my_function`), we would have to always write `pandas.my_function` to refer to it, and this gets cumbersome. A common way to abbreviate "pandas" is "pd", which is why we imported it the way we did above. This way, we can refer to functions in the `pandas` package like `pd.my_function` instead.

Real world data is often messy and needs a bit of pre-processing or "cleaning" before it can by used (easily) for analysis. There is no well-defined list of steps to go through when cleaning data, because it depends on what your raw data looks like to begin with.

The data we'll work with now is a set of makeup reviews from the Sephora online store. The original data can be found [here](https://github.com/everestpipkin/datagardens/tree/master/students/khanniie/5_newDataSet), though we've manually altered the data to help illustrate some data cleaning principles.

### Data cleaning plan

Before we roll up our sleeves and get to cleaning the data, we need to devise a plan:

**Step 1:** *Figure out the start point* - what does our (raw) data look like now?

**Step 2:** *Set analysis goals* - what kind of analyses might we want to run based on the available data?

**Step 3:** *Define the end point* - what form do we want the data in for analysis?

**Step 4:** *Make a task list* - what do we need to do to get our data from the start to the end point?

Repeat these steps as you come up with more analysis ideas! It won't necessarily happen linearly.

Whenever I'm cleaning data, I like to have the `pandas` [cheat sheet](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf) open, because I need to reference it embarassingly often :)

---

### Problem 5 (4 pts)

Perform step 1 in the data cleaning plan by loading the data file (`all_reviews.csv`) using `panda`'s `read_csv` function. Save the data to a variable called `raw_reviews` and display the first few rows of the of the data using the [`.head()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html) data frame method.

(*Remember:* The syntax for object-specific methods looks like `variable_name.method_name()`.)

---

In [25]:
raw_reviews = pd.read_csv("all_reviews.csv")
raw_reviews.head()
raw_reviews.columns

Index(['Brand', 'Product Name', 'Product Type', 'Review Body', 'Review Date',
       'Review Stars', 'Review Title', 'URL', 'User ID'],
      dtype='object')

---

(This data is already a lot cleaner than you might see in practice, we want to start slow!)

**Step 2:** Based on the data that's available, here are some concrete analysis goals (to start!) for this data set:

1. compare ratings for products of the same type
1. look at the trend in ratings over time for a specific product

**Step 3:** So what do we need the clean data to look like for the above analysis? Let's sketch the header for our dream table:

|product_type|brand|product_name|review_stars|review_date|
|------------|-----|------------|------------|-----------|
|str|str|str|int|date|

You'll notice we've written our desired column names as lowercase strings with underscores instead of spaces. Lowercase is easier to type when programming, and spaces in column names prevent us from using some syntax to quickly refer to columns. 

**Step 4:** Now how do we get from the raw data to clean data? It helps to split the tasks into two parts: cleaning column names and cleaning entries.

Let's first clean up the column names and then take a look at the entries to see what needs to be done there. 

### Task List 1: Column  names
1. convert column names to lowercase and replace spaces with underscores
1. drop unnecessary columns

---

### Problem 6 (4 pts)

Use a list comprehension to create a list called `new_colnames` that takes the old column names (stored in `raw_reviews.columns`) and converts them to lowercase. Print `new_colnames`.

*(**Hint:** You may find the string method [`.lower()`](https://www.w3schools.com/python/ref_string_lower.asp) helpful here)*

---

In [16]:
new_colname = [column_name.lower() for column_name in raw_reviews.columns]
print(new_colname)

['brand', 'product name', 'product type', 'review body', 'review date', 'review stars', 'review title', 'url', 'user id']


---

### Problem 7 (8 pts)

Use another list comprehension update the `new_colnames` list, replacing all spaces in the list entries with underscores. Print the updated `new_colnames` list.

Copy `raw_reviews` to a new variable called `reviews` using the [`.copy()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.copy.html) data frame method. Update the column names for the `reviews` data frame using your cleaned column names stored in `new_colnames`. Show the last few rows of `reviews`.

*(**Hint:** you may find the string method [`.replace()`](https://www.w3schools.com/python/ref_string_replace.asp) helpful here)*

#### Task List 2: Column  names cont.
1. convert spaces in column names to '_' and print the updated `new_colnames` list
1. copy the contents of `raw_reviews` to a new data frame called `reviews`
1. display the last few rows of `reviews`
---

In [41]:
new_colname = [col.replace(" ", "_") for col in new_colname]
reviews = raw_reviews.copy()
reviews.columns =  new_colname

reviews.tail()

brand                                   product_name  \
3961              Urban Decay             Eyeshadow Primer Potion - Original   
3962              Urban Decay  All Nighter Long-Lasting Makeup Setting Spray   
3963              Urban Decay  All Nighter Long-Lasting Makeup Setting Spray   
3964  FENTY BEAUTY BY RIHANNA              Flyliner Longwear Liquid Eyeliner   
3965  FENTY BEAUTY BY RIHANNA              Flyliner Longwear Liquid Eyeliner   

         product_type                                        review_body  \
3961       Eye Primer  I have gone to bed many of nights after drinki...   
3962    Setting Spray  At first, I thought makeup setting sprays were...   
3963    Setting Spray  This product is great at keeping makeup on for...   
3964  Liquid Eyeliner  I really wanted to love this liner. So many re...   
3965  Liquid Eyeliner                      Too much transfer onto eyelid   

     review_date review_stars  \
3961  31 12 2015      5 stars   
3962  31 12 2015      5 stars   
3963  31 12 2015      5 stars   
3964  31 12 2018      2 stars   
3965  31 12 2018      3 stars   

                                           review_title  \
3961                                         Holy Grail   
3962      skeptical at first, but it made me a believer   
3963                                I always repurchase   
3964  <img src="https://www.sephora.com/productimage...   
3965  <img src="https://www.sephora.com/productimage...   

                                                    url       user_id  
3961  https://www.sephora.com/product/eyeshadow-prim... -8.830000e+18  
3962  https://www.sephora.com/product/all-nighter-lo... -1.160000e+18  
3963  https://www.sephora.com/product/all-nighter-lo...  6.990000e+18  
3964  https://www.sephora.com/product/flyliner-longw... -3.430000e+18  
3965  https://www.sephora.com/product/flyliner-longw... -3.850000e+18

---

### Problem 8 (4 pts)

Update the `reviews` data frame, keeping only the following columns:

```
 brand, product_type, review_stars, review_date
```
in this specified order. 

Show the first 3 rows of the updated `reviews` data frame.

---

In [44]:
reviews = reviews[["brand", "product_type", "review_stars", "review_date"]]
reviews.head(3)

brand     product_type review_stars review_date
0         Dior          Mascara      5 stars  01 01 2011
1    Kat Von D  Liquid Eyeliner      5 stars  01 01 2011
2  Urban Decay          Mascara       1 star  01 01 2012

---

Now that our column names are in order, let's look at what kind of cleaning our entries need.

---
### Problem 9 (6 pts)

Write a for loop that will, for the columns `brand`, `product_type`, `review_stars` in `reviews`,
1. print the column name.
1. print the count of observations for the type with the most entries in the given column in ascending order.
1. print the 5th entry in the column of the original data (not the aggregated list of value counts).
1. print an empty line break (`"\n"`) to make the results produced by the for loop more readible.

*(**Hint:** For step 2 in the for loop, you may find the `pandas` column method [`.value_counts()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html?highlight=value_counts#pandas.Series.value_counts) helpful.)*

---

In [71]:
for reviews.col in reviews.columns:
    print(reviews.col)
    print(reviews.col.value_counts())
    print(reviews.col[:,5])

brand


AttributeError: 'str' object has no attribute 'value_counts'

---

### Problem 10 (4 pts)

Look at the results from the previous exercise for the `review_stars` column. What do you notice about some of the entries that might be a problem for data analysis? Give a way that we could fix this issue. 

---

## insert answer here

---

Inspecting the entries, we should notice a few things that may be a problem for our analysis plan:
* some entries in the `product_type` column are inconsistent (e.g. `Mascara` and `MASCARA`)
* there appear to be some observations with missing values in the `review_stars` column (some observations have `???` in this column)
* the dates don't really look like dates and may not be parsed correctly
    
Using these observations, we can write our second task list for cleaning the table entries.

#### Task List 2: Entries
1. convert entries in the `product_type` column to a similar format, to get rid of duplicate categories arrising from case differences
1. remove observations with missing values in the `review_stars` column
1. convert the `review_dates` column to an actual `pandas` date object

---

### Problem 11 (4 pts)

Update `reviews` by converting the `product_type` column to only have entries where the first letter of each entry is capitalized. Display the first few rows of the updated `reviews`.

When updating a subset of a `pandas` dataframe, be sure to use the [`.loc[]`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html?highlight=loc#pandas.DataFrame.loc) data frame method on the left-hand side of the assignment statement. You should **not** get a warning labelled `SettingWithCopyWarning` if you've done this correctly.

---

In [5]:
## insert answer here

---

### Problem 12 (6 pts)

Print the rows of `reviews` where the `review_stars` column has the entry '???'.

Then, update reviews by removing any rows where the review_stars entry is ???. Display the first few rows of the updated reviews.

*(**Confidence check:** Your output should contain 2 rows of the `reviews` dataframe. The new dataframe should also include 3964 entries.*

---

In [85]:
rows_to_drop = reviews.loc[reviews.review_stars == "???"]
rows_to_drop[0]
#reviews.drop(index = [458, 3400])

pandas.core.frame.DataFrame

---

Note that often missing values will appear as null values (as opposed to something like `???`). Null values depend on the underlying data type: `NaN` in numeric arrays, `None` or `NaN` in object arrays, `NaT` in datetime arrays. The functions [`isnull()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.isnull.html) and [`notnull()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.notnull.html?highlight=notnull#pandas.DataFrame.notnull) are helpful for filtering in these cases.  You should **not** use logical operators like `==` or `!=` to check for null values, as such expressions often have [unintended consequences](https://towardsdatascience.com/navigating-the-hell-of-nans-in-python-71b12558895b).

That said, it is worth asking yourself why you're removing rows with missing values and if it might affect some downstream analysis. For example, if we were interested in the *number* of reviews per product, removing products with missing ratings would artificially reduce these sample sizes.

Now, we just have to tackle the dates and we will have accomplished all of our data cleaning tasks!

Let's take a closer look at the dates, as they currently appear in `reviews`:

In [15]:
type(reviews.review_date[0])

str

The review dates are currently stored as strings in `reviews`. Python does not inately know a special date string from any other string (functionally, it thinks of "2020-09-07" and "hello" as the same type of data). Information about how dates should be read, ordered, and how to perform calculations with dates is not encoded when a date is stored as a string.

Instead, we must represent dates as specific "date" types in Python. To do this, we have to tell Python how to parse the date string to create a special date object. 

---

### Problem 13 (6 pts)

Update `reviews` by overwriting the exisiting string entries in the `review_dates` column with correctly parsed date entries where the day is written first. Use the `pandas` [`to_datetime()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html) function to convert the dates in the existing `review_date` column to ones that are correctly parsed using the `format` argument.

Display the first 6 rows of the updated `reviews`. Print the data type of the 1st entry of the updated `review_date` column.

*(**Hint:** The original dates in the `review_date` column are of the form `dd mm yyyy`. In writing the `format` string that parses this formatting, you may find it helpful to consult [this chart](https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes) of date and time codes recognized by `format`. For example, `format = "%b %d, %y"` would capture dates like `Jul 15, 91`.)*

#### Task List
1. convert the `review_date` column to correctly parsed date entries where the day is written first. 
1. display the first 6 rows of the updated `reviews`
1. print the data type of the 1st entry of the updated `review_date` column.
---

In [5]:
## insert answer here

---

That's it! We've accomplished our data cleaning task list. Of course, as you begin to work with your data, you may notice other idiosyncrasies that require further cleaning.

For instance, the entries in the `review_stars` column are still strings, but in order to perform certain kinds of analyses (e.g. get the mean star rating per item), we really want this data to be stored as a numeric type... How might one use Python to extract the numeric value of the rating? More on this later, when we start learning about [regular expressions](https://www.programiz.com/python-programming/regex)... 